### PR TITLE
feat: managed object storage support

### PR DESCRIPTION
**Pull request by villevsv-upcloud**

### CODE REVIEW

File: internal/service/managedobjectstorage/managed_object_storage.go [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+			"endpoint": {
+				Description: "Endpoints for accessing the Managed Object Storage service.",
+				Computed:    true,
+				Type:        schema.TypeSet,
+				Elem:        schemaEndpoint(),
+			},
+			"labels": utils.LabelsSchema("network gateway"),
```
Comment (kangasta) [inline internal/service/managedobjectstorage/managed_object_storage.go:53]: `network gateway` → `managed object storage`

File: internal/service/managedobjectstorage/managed_object_storage.go [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+				Type:        schema.TypeSet,
+				Elem:        schemaNetwork(),
+			},
+			"operational_state": {
+				Description: "Operational state of the Managed Object Storage service.",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+			"region": {
+				Description: "Region in which the service will be hosted, see `upcloud_object_storage2_regions` data source.",
```
Comment (kangasta) [inline internal/service/managedobjectstorage/managed_object_storage.go:66]: `upcloud_object_storage2_regions` → `upcloud_managed_object_storage_regions`

File: internal/service/managedobjectstorage/managed_object_storage.go [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+				Required:    true,
+				Type:        schema.TypeString,
+			},
+			"updated_at": {
+				Description: "Creation time.",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+			"users": {
+				Description: "List of cloud Users allowed to use object storage. See `upcloud_managed_object_storage_user_access_key` for managing access keys.",
```
Comment (kangasta) [inline internal/service/managedobjectstorage/managed_object_storage.go:76]: Does _cloud user_ mean main-account and/or sub-account here or is this related to a DB resource?

File: internal/service/managedobjectstorage/managed_object_storage.go [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+					Type: schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+func schemaEndpoint() *schema.Resource {
+	return &schema.Resource{
+		Description: "OpenSearch index",
```
Comment (kangasta) [inline internal/service/managedobjectstorage/managed_object_storage.go:89]: `OpenSearch index` → `Endpoint`

File: internal/service/managedobjectstorage/managed_object_storage.go [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+		Schema: map[string]*schema.Schema{
+			"domain_name": {
+				Description: "Domain name of the endpoint.",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+			"type": {
+				Description: "Type of the endpoint (`private` / `public`).",
+				Computed:    true,
+				Type:        schema.TypeString,
```
Comment (kangasta) [inline internal/service/managedobjectstorage/managed_object_storage.go:99]: Could add validation here, e.g.: 

```suggestion
				Type:        schema.TypeString,
				ValidateFunc: validation.StringInSlice([]string{"private", "public"}, false),
```

File: internal/service/managedobjectstorage/managed_object_storage.go [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+				ValidateFunc: validation.StringInSlice(
+					[]string{
+						"private",
+						"public",
+					},
+					false,
+				),
+			},
+			"uuid": {
```
Comment (kangasta) [inline internal/service/managedobjectstorage/managed_object_storage.go:142]: LBaaS uses `network` instead of `uuid` even though both have `uuid` in API. Otherwise the network sub-schema seems to be equal.

File: internal/service/managedobjectstorage/managed_object_storage.go [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+	req.Users = userReqs
+
+	storage, err := svc.CreateManagedObjectStorage(ctx, req)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(storage.UUID)
+
+	waitReq := &request.WaitForManagedObjectStorageOperationalStateRequest{
```
Comment (kangasta) [inline internal/service/managedobjectstorage/managed_object_storage.go:188]: Not related to this PR, but should we pass context instead of timeout to these `WaitFor*State` functions?

File: internal/service/managedobjectstorage/managed_object_storage.go [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+	}
+
+	storage, err = svc.WaitForManagedObjectStorageOperationalState(ctx, waitReq)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return append(diags, setManagedObjectStorageData(d, storage)...)
+}
+
```
Comment (kangasta) [inline internal/service/managedobjectstorage/managed_object_storage.go:205]: I'll continue reviewing from here tomorrow

File: internal/service/managedobjectstorage/managed_object_storage.go (hunk 7 above)
Comment (villevsv-upcloud) [reply internal/service/managedobjectstorage/managed_object_storage.go:188]: Good point, ticket created.

File: internal/service/managedobjectstorage/user_access_key.go [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+	if accessKey.SecretAccessKey != nil {
+		if err := d.Set("secret_access_key", accessKey.SecretAccessKey); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	return diags
+}
+
+func marshalID(components ...string) string {
```
Comment (kangasta) [inline internal/service/managedobjectstorage/user_access_key.go:201]: Do/should we have common helper for this? I think some other components use similar combined key as well

File: internal/service/managedobjectstorage/managed_object_storage.go (hunk 3 above)
Comment (villevsv-upcloud) [reply internal/service/managedobjectstorage/managed_object_storage.go]: Current main account and it's sub-accounts

File: internal/service/managedobjectstorage/managed_object_storage.go (hunk 6 above)
Comment (villevsv-upcloud) [reply internal/service/managedobjectstorage/managed_object_storage.go:142]: Let's keep this as `uuid`.

File: internal/service/managedobjectstorage/managed_object_storage.go (hunk 5 above)
Comment (villevsv-upcloud) [reply internal/service/managedobjectstorage/managed_object_storage.go:99]: Added, though not strictly necessary for a computed field.

File: internal/service/managedobjectstorage/user_access_key.go (hunk 9 above)
Comment (villevsv-upcloud) [reply internal/service/managedobjectstorage/user_access_key.go:201]: Added [a separate PR](https://github.com/UpCloudLtd/terraform-provider-upcloud/pull/420).

File: internal/service/managedobjectstorage/managed_object_storage.go (hunk 5 above)
Comment (kangasta) [reply internal/service/managedobjectstorage/managed_object_storage.go:99]: Ah, sorry, didn't notice the `Computed: true`, seems that `ValidateFunc` cannot be used in this case

Comment (villevsv-upcloud) [reply internal/service/managedobjectstorage/managed_object_storage.go:99]: No worries, it was useless to add it in any case. :) Reverted the change just now.

**Review by kangasta — APPROVED** [review]
